docs/how-to/push-update-on-mainnet.md

```diff
@@ -0,0 +1,120 @@
+3. Software update is confirmed by voting from majority of nodes.
+4. Nodes that see an update try to download and apply it
+   automatically.
+
+Now, again, step-by-step.
+
+Prerequisites
+=============
+1. Cluster nodes' keys. They are crucial to have since cluster nodes
+   are the only nodes that have stake in the system so they are the
+   only nodes that can propose an update and vote for it.
+2. Access to the S3 bucket
+3. Installers corresponding to software version we're pushing to cluster
+   - Those are two executables (exe for win, pkg for mac) which are to be provided by QA (after QA procedures passed)
+   - Advice: after providing installers, ask QA to confirm hashes are same, we don't want to ship users incorrect installers to users
+4. Software version
+   - Single integer, denote application version as for installers provided
+
+Proposing the update
+====================
+
+Ensure your keys are somewhere around and named, say, `nodeX.key`,
+where `X ∈ [1..7]`. Procedure is done using `cardano-auxx`
+executable.
+
+You should change param `--peer` for all commands listed below. Also consider changing `--log-config`, `--logs-prefix`.
+Note `--system-start 0` is perfectly valid because we don't need actual value for `cardano-auxx`.
+
+Import secret keys: only 4 of 7 (majority) is needed.
+
+```
+stack exec cardano-auxx -- --system-start 0 --log-config
+  scripts/log-templates/log-config-qa.yaml
+  --logs-prefix "logs/auxx-update-1.0.1" --db-path auxx-update-1.0.1
+  --peer <relay dns name>:3000 cmd --commands "add-key node1.key, add-key node2.key, add-key node3.key, add-key node4.key, listaddr"
+```
+
+Propose an update:
+
+```
+stack exec cardano-auxx -- --system-start 0 --log-config
+  scripts/log-templates/log-config-qa.yaml
+  --logs-prefix "logs/auxx-update-1.0.1" --db-path auxx-update-1.0.1
+  --peer <relay dns name>:3000 cmd --commands "propose-update 0 0.0.0 0 20 2000000 csl-daedalus:1 win64 daedalus1.exe none macos daedalus1c.pkg none"
+```
+
+Syntax and semantics of `propose-update` command:
+
+```
+propose-update <N> <block ver> <script ver> <slot duration> <max block size> <software ver> <propose_file>?
+```
+
+* First argument is index of imported key you're sending update
+from. Step "import secret keys" has `listaddr` command in the end. Check it's output -- any cluster node will
+do (`0` states for first key imported).
+* Parameters 2-5 are block version data parameters -- leave them as
+they are provided in the cli example.
+* `csl-daedalus:1` is software version description, you should substitute `1` (version) with the integer provided along with installers (see *Prerequisites* section, *4.* item)
+* End of command: list of triples -- update installers. Provide the path to installers
+without any slashes (installers should be in the same dir as auxx is
+launched in). First tuple element is platform, second is path, `none` stands for binary diff (this feature is not used for now).
```

Comment:
Is issuing an update to only one platform (`win64` or `macos`) supported? If not, does `cardano-auxx` require one triple tuple for each?

Reply:
> does cardano-auxx require one triple tuple for each [platform]?

Yes. `propose-update ... <platform1 installerPath1 diffPath1> <platform2 installerPath2 diffPath2> ...`

> Is issuing an update to only one platform (win64 or macos) supported?

Yes.

Reply:
What about the version gap for the non-updated platform, @neongreen?  The scenario:

1. Suppose both platforms A and B start with version N.
2. Update round 1: Platform A is updated to N+1, platform B doesn't get an update
3. Update round 2: both platforms are updated

What are the supposed version numbers in this case? I presume N+2 and N+1.  But that means we have completely independent versioning for the platforms.

Can you foresee any problems with that? /cc @akegalj

Reply:
If you update platform A to N+1, you won't be able to update B to N+1 later – you'll need to update both to N+2.

* You publish an update to N+1 that contains an installer for Windows. N+1 becomes a confirmed version after `k` blocks. Windows nodes download the update. macOS nodes don't download anything.
* Someone else publishes an update to N+5. It gets rejected because the current confirmed version is N+1.
* You publish an update to N+2 that contains both a Windows and a macOS installer. N+2 becomes a confirmed version after `k` blocks. Windows and macOS nodes download the update (even though for Windows it's N+1→N+2 and for macOS it's N→N+2).

Reply:
> Can you foresee any problems with that?

If the Daedalus installers don't rely on some particular old files being around – no, there shouldn't be any problems.

Reply:
What @neongreen just said contradicts the previous requrement, where version names are monotonically increasing.

So we have an issue here that should be documented: all platforms should be updated at the same time to the next version.

Reply:
I'm highly confused :(

>where version names are monotonically increasing.

What? Version names for each platform will be montonically increasing, though non-consequent versions are possible.

>So we have an issue here that should be documented: all platforms should be updated at the same time to the next version.

How did you come to this statement?

Reply:
Application versions must be consequent. Blockchain will reject an update to version 3 if the last confirmed version is 1.
Regarding installers, nothing is checked by blockchain protocol. We can include arbitrary number of installer hashes into update proposal (well, there is upper limit on proposal size), we can use arbitrary system tags, etc. It's up to us to decide what which installers we want to include.

> So we have an issue here that should be documented: all platforms should be updated at the same time to the next version.

If you mean that we must always include all supported system tags into update proposal, then it's wrong. We can update only Windows or not update anything. And it might be useful in case we fix only Windows-specific bug.
However, it we want to update all platforms to version X, then we must include all system tags into update proposal. Updating to X more than once is impossible.

Reply:
It's essential to understand that application version is associated with application name (e. g. `csl-daedalus`), not with system tag (e. g. `win64`).
Application version is checked by protocol. Installer hashes are not checked in any way.